---
title: "Hawaiian Soil Microbiome"
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: united
---

### Load Libraries and Functions
```{r, include=FALSE}
library(arules)
library(phyloseq)
library(ggplot2)
library(dplyr)
library(vegan)
library(stringr)
library(DESeq2)
library(EMA)
library(reshape2)
```

```{r warning=FALSE, echo=FALSE}
library(phyloseq)
library(ggplot2)
library(dplyr)
library(vegan)
library(stringr)
library(DESeq2)
library(EMA)
library(reshape2)
library(extrafont)
#font_import()
```

```{r, echo=FALSE}
## Helper Functions

### Confidence Intervals
ci = function(x){mean(x)+c(-1.96,1.96)*sd(x)/sqrt(length(x))}

### Correlation
my_fun<-function(x){
  vals = c(vals, cor.test(x, level)[['p.value']])
}

my_plot_bar = function (physeq, x = "Sample", y = "Abundance", fill = NULL, title = NULL, facet_grid = NULL) {
    mdf = psmelt(physeq)
    p = ggplot(mdf, aes_string(x = x, y = y, fill = fill))
    p = p + geom_bar(stat = "identity", position = "stack", show.legend = T)
    p = p + theme(axis.text.x = element_text(angle = -90, hjust = 0))
    if (!is.null(facet_grid)) {
        p <- p + facet_grid(facet_grid)
    }
    if (!is.null(title)) {
        p <- p + ggtitle(title)
    }
    return(p)
}

```


### Load Data and Create Phyloseq Objects
```{r warning=FALSE, echo=FALSE}
load("../data/DADA2_Processed/Taxa_Silva_8_18_Hawaiian.RData")
load("../data/DADA2_Processed/seqtabnochim_8_18_Hawaiian.RData")
labels = read.csv("../data/DADA2_Processed/NTM_Count_Bi.csv")

## Remove samples not in soil chemistry
labels_to_remove = c("12_56", "12_7")
labels = subset(labels, !(SampleName %in% labels_to_remove))

## Remove Samples not in Soil Chemistry HNL 12-56 & HNL 12-7 
rows_to_remove = c("HIsoil04", "HIsoil09", "HIsoil00")
seqtab.nochim = subset(seqtab.nochim, !(rownames(seqtab.nochim) %in% rows_to_remove))
empties = (colSums(seqtab.nochim) != 0)
seqtab.nochim = seqtab.nochim[, empties]

## Load Metadata and edit naming
samdf = read.csv(file = "../data/DADA2_Processed/metadata.txt", sep='\t', header = T, stringsAsFactors = FALSE)
row.names(samdf) = samdf$X.SampleID
samdf = subset(samdf, !(rownames(samdf) %in% rows_to_remove))
rownames(labels) = labels$SampleName
labels$SampleName = NULL
samdf = cbind(samdf, labels)
samdf$SampleName = paste('HNL ',samdf$SampleName, sep = "")
samdf$SampleName = lapply(samdf$SampleName, function(x){gsub("_", "-", x)})

## Filter Tax Table
x = colnames(seqtab.nochim)
taxa_silva = taxa_silva[rownames(taxa_silva) %in% x,]

### Load in Soil characterisitics
df = read.csv("../data/DADA2_Processed/Soil_Scaled.csv")
rownames(df) = df$X
df$X = NULL

## Create Phyloseq Object and rarified Phyloseq Object
ps = phyloseq(otu_table(seqtab.nochim,taxa_are_rows = FALSE),sample_data(samdf), tax_table(taxa_silva))
set.seed(1)
ps_rarified = rarefy_even_depth(ps,sample.size = 9214, verbose = F)
ps_phylum = tax_glom(ps_rarified, taxrank = "Phylum")
ps_family = tax_glom(ps_rarified, taxrank = "Family")
```


### Correlate OTU with Outcome Variable
Using DeSeq2 (Negative Binomial Correlation), only one OTU is differentially expressed between NTM culture status, Actinomycetospora. 
```{r warning=FALSE, echo=FALSE}
# ### Full Dataset
# diagdds = phyloseq_to_deseq2(ps_rarified, ~NTM_culture)
# gm_mean = function(x, na.rm=TRUE){
#   exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
# }
# 
# geoMeans = apply(counts(diagdds), 1, gm_mean)
# diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
# diagdds = DESeq(diagdds, fitType="local")
# 
# res = results(diagdds)
# res = res[order(res$padj, na.last=NA), ]
# alpha = 0.01
# sigtab = as.data.frame(res[(res$padj < alpha), ])
# Full_sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_rarified)[rownames(sigtab), ], "matrix"))
# 
# ### Grouping by genus does not have any significantly different groups
# 
# ### Family
# diagdds = phyloseq_to_deseq2(ps_family, ~NTM_culture)
# gm_mean = function(x, na.rm=TRUE){
#   exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
# }
# 
# geoMeans = apply(counts(diagdds), 1, gm_mean)
# diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
# diagdds = DESeq(diagdds, fitType="local")
# 
# res = results(diagdds)
# res = res[order(res$padj, na.last=NA), ]
# alpha = 0.01
# sigtab = as.data.frame(res[(res$padj < alpha), ])
# Family_sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_rarified)[rownames(sigtab), ], "matrix"))
# 
# ### Phylum
diagdds = phyloseq_to_deseq2(ps_phylum, ~NTM_culture)
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

geoMeans = apply(counts(diagdds), 1, gm_mean)
diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
diagdds = DESeq(diagdds, fitType="local")

res = results(diagdds)
res = res[order(res$padj, na.last=NA), ]
alpha = 0.01
sigtab = as.data.frame(res[(res$padj < alpha), ])
Phylum_sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_rarified)[rownames(sigtab), ], "matrix"))
```

### Plot Significant OTUS

#### Full OTUs
##### Actinomycetospora
```{r, echo=FALSE}
Actino = subset_taxa(ps_rarified, Genus=="Actinomycetospora")
# png('../Images/SupplementaryActinomyetospora.png', units="in", width=6, height=8, res=600)
# p = plot_bar(Actino, x="NTM_culture", fill = "NTM_culture")
# 
# p + theme_bw() + geom_bar(aes(color=NTM_culture, fill=NTM_culture), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = "Actinomycetospora") + labs(title="Total Actinomycetospora Counts by NTM Culture Status", family="Arial") + scale_x_discrete(labels= c("Negative\nCulture", "Positive\nCulture")) + theme(axis.text=element_text(size=18),axis.title=element_text(size=14,face="bold", family="Arial"), axis.text.x = element_text(angle = 0, hjust = .5, face="bold", family = "Arial", size = 16)) + ylab("Read Abundance") + xlab("") 
# dev.off()
```

#### Family Level
##### Pasteurellaceae
```{r, echo=FALSE}
Pasteurellaceae = subset_taxa(ps_family, Family=="Pasteurellaceae")

# png('../Images/SupplementaryPasteurellaceae.png', units="in", width=6, height=8, res=1200)
# p = plot_bar(Pasteurellaceae, x="NTM_culture", fill = "NTM_culture")
# 
# p + theme_bw() + geom_bar(aes(color=NTM_culture, fill=NTM_culture), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = "Pasteurellaceae") + labs(title="Total Pasteurellaceae Counts by NTM Culture Status", family="Arial") + scale_x_discrete(labels= c("Negative\nCulture", "Positive\nCulture")) + theme(axis.text=element_text(size=18),axis.title=element_text(size=14,face="bold", family="Arial"), axis.text.x = element_text(angle = 0, hjust = .5, face="bold", family = "Arial", size = 16)) + ylab("Read Abundance") + xlab("")  
# dev.off()
```

##### Sporolactobacillaceae
```{r, echo=FALSE}
Sporolactobacillaceae = subset_taxa(ps_family, Family=="Sporolactobacillaceae")

# png('../Images/SuuplementarySporolactobacillaceae.png', units="in", width=6, height=8, res=1200)
# p = plot_bar(Sporolactobacillaceae, x="NTM_culture", fill = "NTM_culture")
# 
# p + theme_bw() + geom_bar(aes(color=NTM_culture, fill=NTM_culture), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = "Sporolactobacillaceae") + labs(title="Total Sporolactobacillaceae Counts by NTM Culture Status", family="Arial") + scale_x_discrete(labels= c("Negative\nCulture", "Positive\nCulture")) + theme(axis.text=element_text(size=18),axis.title=element_text(size=14,face="bold", family="Arial"), axis.text.x = element_text(angle = 0, hjust = .5, face="bold", family = "Arial", size = 16)) + ylab("Read Abundance") + xlab("")
# dev.off()
```

#### Phylum Level
##### Firicutes
```{r, echo=FALSE}
Firmicutes = subset_taxa(ps_phylum, Phylum=="Firmicutes")

png('../Images/SupplementaryFigure2B.png', units="in", width=6, height=8, res=1200)
p = plot_bar(Firmicutes, x="NTM_culture", fill = "NTM_culture")

p + theme_bw() + geom_bar(aes(color=NTM_culture, fill=NTM_culture), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = "Firmicutes") + labs(title="Total Firmicutes Counts by NTM Culture Status", family="Arial") + scale_x_discrete(labels= c("Negative\nCulture", "Positive\nCulture")) + theme(axis.text=element_text(size=18),axis.title=element_text(size=14,face="bold", family="Arial"), axis.text.x = element_text(angle = 0, hjust = .5, face="bold", family = "Arial", size = 16)) + ylab("Read Abundance") + xlab("")
dev.off()
```




### Alpha Diversity
#### NTM Culture
```{r, echo=FALSE}
erich <- estimate_richness(ps_rarified, measures = c("Observed", "Shannon", "Simpson", "Fisher"))
ttest_culture <- t(sapply(erich, function(x) unlist(t.test(x~sample_data(ps_rarified)$NTM_culture)[c("estimate","p.value","statistic","conf.int")])))

erich["X.SampleID"] = rownames(erich)
erich2 = as.data.frame(erich)
x = sample_data(ps_rarified)
names = x$SampleName
names = unlist(names)
erich["SampleName"] = names
colsto = c("SampleName", "Shannon")
y2 = erich[colnames(erich) %in% colsto]
write.csv(y2, "../data/DADA2_Processed/ShannonValues.csv", quote = F, row.names = F)

erich <- estimate_richness(ps_rarified, measures = c("Observed", "Shannon", "Simpson", "Fisher"))
ttest_count <- t(sapply(erich, function(x) unlist(t.test(x~sample_data(ps_rarified)$NTM_Counts_Categories)[c("estimate","p.value","statistic","conf.int")])))

write.csv(ttest_culture, "../data/DADA2_Processed/ttest/Culture.csv", quote = F, row.names = T)
write.csv(ttest_count, "../data/DADA2_Processed/ttest/Counts.csv", quote = F, row.names = T)
```

```{r fig.width=12, fig.height=6, echo=FALSE}
x = estimate_richness(ps_rarified, measures = c("Shannon"))
samdf = ps_rarified@sam_data
vac_dust = samdf[, c("NTM_culture", "NTM_Counts_Categories")]
vac = merge(x, vac_dust, by = 0)

## Change Culture Status
nos = vac[vac$NTM_culture == "negative",]
yes = vac[vac$NTM_culture == "positive",]

ci_no = ci(nos$Shannon)
ci_yes = ci(yes$Shannon)
mean_no = mean(nos$Shannon)
no_value = mean(nos$Shannon)*sd(nos$Shannon)/sqrt(length(nos$Shannon))
paste(as.character(mean_no), " +/- ", as.character(no_value), sep = "")
mean_yes = mean(yes$Shannon)
yes_value = mean(yes$Shannon)*sd(yes$Shannon)/sqrt(length(yes$Shannon))
paste(as.character(mean_yes), " +/- ", as.character(yes_value), sep = "")
nos_vec = as.data.frame(c(mean_no, ci_no))
yes_vec = c(mean_yes, ci_yes)
nos_vec = cbind(nos_vec, yes_vec)

## Change Group
colnames(nos_vec) = c("Negative", "Positive")

nos_vec = as.data.frame(t(nos_vec))
colnames(nos_vec) = c("Mean", "Lower_CI", "Upper_CI")
nos_vec$Group = rownames(nos_vec)
nos_vec$Feature = "NTM Culture Status"
nos_melted = melt(nos_vec, "Group")

total_cis = nos_vec

png('../Images/SupplementaryFigure2A.png', units="in", width=5, height=8, res=600)
limits <- aes(ymax = total_cis$Upper_CI,ymin = total_cis$Lower_CI)

p <- ggplot(data = total_cis, aes(x = Feature, y = Mean, fill = Group))
p + geom_bar(stat = "identity",position = position_dodge(0.9), show.legend = F) + geom_errorbar(limits, position = position_dodge(0.9),width = 0.25) + geom_text(aes(Feature,-.1, label = Group), position = position_dodge2(width=0.9, padding = 2),  size=6) + scale_fill_manual(values=c("#4863A0", "#FFF380")) + theme_bw() + theme(axis.text.y = element_text(colour="grey20",size=18,angle=0,hjust=1,vjust=0,face="plain", family = "Arial"),
                                                                      axis.title.y = element_text(colour="grey20",size=18,angle=90,hjust=.5,vjust=.5,face="plain", family = "Arial"),
                                                                      axis.text.x = element_text(colour="grey20",size=18,angle=0,hjust=0.5,vjust=0,face="plain", family = "Arial"),
                                                                      axis.title.x = element_text(colour="grey20",size=18,angle=0,hjust=.5,vjust=.5,face="plain", family = "Arial")) + xlab("") + ylab("Shannon Diversity Index") +ggtitle("Alpha Diversity by Mycobacterium Culture Status")
dev.off()
```

#### NTM Count
```{r, echo=FALSE}
x = estimate_richness(ps_rarified, measures = c("Shannon"))
samdf = ps_rarified@sam_data
vac_dust = samdf[, c("NTM_culture", "NTM_Counts_Categories")]
vac = merge(x, vac_dust, by = 0)

## Change NTM Status
nos = vac[vac$NTM_Counts_Categories == "Low",]
yes = vac[vac$NTM_Counts_Categories == "High",]


ci_no = ci(nos$Shannon)
ci_yes = ci(yes$Shannon)
mean_no = mean(nos$Shannon)
no_value = mean(nos$Shannon)*sd(nos$Shannon)/sqrt(length(nos$Shannon))
paste(as.character(mean_no), " +/- ", as.character(no_value), sep = "")
mean_yes = mean(yes$Shannon)
yes_value = mean(yes$Shannon)*sd(yes$Shannon)/sqrt(length(yes$Shannon))
paste(as.character(mean_yes), " +/- ", as.character(yes_value), sep = "")
nos_vec = as.data.frame(c(mean_no, ci_no))
yes_vec = c(mean_yes, ci_yes)
nos_vec = cbind(nos_vec, yes_vec)
colnames(nos_vec) = c("Low", "High")

nos_vec = as.data.frame(t(nos_vec))
colnames(nos_vec) = c("Mean", "Lower_CI", "Upper_CI")
nos_vec$Group = rownames(nos_vec)
nos_vec$Feature = "Discretized NTM Microbiome Counts"
nos_melted = melt(nos_vec, "Group")

total_cis = rbind(total_cis, nos_vec)
total_cis = nos_vec


#png('../Images/AlphaDiversitybyCounts.png', units="in", width=5, height=8, res=600)
# limits <- aes(ymax = total_cis$Upper_CI,ymin = total_cis$Lower_CI)
# 
# p <- ggplot(data = total_cis, aes(x = Feature, y = Mean, fill = Group))
# p + geom_bar(stat = "identity",position = position_dodge(0.9), show.legend = F) + geom_errorbar(limits, position = position_dodge(0.9),width = 0.25) + geom_text(aes(Feature,-.1, label = Group), position = position_dodge2(width=0.9, padding = 2),  size=6) + scale_fill_manual(values=c("#4863A0", "#FFF380")) + theme_bw() + theme(axis.text.y = element_text(colour="grey20",size=18,angle=0,hjust=1,vjust=0,face="plain", family = "Arial"),
#                                                                       axis.title.y = element_text(colour="grey20",size=18,angle=90,hjust=.5,vjust=.5,face="plain", family = "Arial"),
#                                                                       axis.text.x = element_text(colour="grey20",size=18,angle=0,hjust=0.5,vjust=0,face="plain", family = "Arial"),
#                                                                       axis.title.x = element_text(colour="grey20",size=18,angle=0,hjust=.5,vjust=.5,face="plain", family = "Arial")) + xlab("") + ylab("Shannon Diversity Index") +ggtitle("Alpha Diversity by Mycobacterium Count Category")
# dev.off()
```

### Beta Diversity
To explore Beta Diversity, PCoA clusting of the rarified data is performed. Bray-Curtis is the distance metric utilized for the beta diversity analysis. 

#### Beta Diversity by Culture Status and NTM Counts

```{r fig.width=12, fig.height=6, echo=FALSE}
# png('../Images/SupplementaryBrayBetaDiversity.png', units="in", width=8, height=8, res=600)
# ordinal = ordinate(ps_rarified, method = "PCoA", distance = "bray")
# p7 = plot_ordination(ps_rarified, ordinal, type = "samples", color = "NTM_culture", shape = "NTM_Counts_Categories", title = "Beta Diversity by Culture Status")
# p7 + geom_point(size=7, alpha=0.75)
# dev.off()
```

Adonis test rejects the NULL hypothesis that the distance within groups of NTM Culture Status is from the same distribution as the distance between groups. The plot shows the assumption of heterogeneity is upheld providing more support to the Adonis test.  

```{r, echo=FALSE}
set.seed(1)
samdf = sample_data(ps_rarified)

bray_ps = phyloseq::distance(ps_rarified, method = "bray")
adonis(formula = bray_ps ~ samdf$NTM_culture, as(sample_data(ps_rarified), "data.frame"))

beta <- betadisper(bray_ps, samdf$NTM_culture)
beta.HSD <- TukeyHSD(beta)
plot(beta.HSD, las=2)
```



### Taxa Boxplots
The taxa boxplots are derived from classifications of the core variant sequences using the RDP Classifier from the Silva Database. The Phlyum and Genus Levels are presented. The number of samples in the metadata groups are not equally distributed and as such the graphics do not always have the same y axis value. 

#### Phylum Biodiversity
```{r warning=FALSE, echo=FALSE}
ps_n = transform_sample_counts(ps_rarified, function(x) x / sum(x))
b2 = tax_glom(ps_n, taxrank = "Phylum")
Phylum10 = names(sort(taxa_sums(b2), TRUE)[1:6])
x2 = prune_taxa(Phylum10, b2)
## Add Grouped Column
x = as.data.frame(otu_table(x2))
x$Other = 1 - rowSums(x)
x = as.matrix(x)

q = tax_table(x2)
c = q@.Data
taxax = as.data.frame(c)
Other = data.frame(Kingdom="Other", Phylum="Other", Class=NA, Order=NA, Family=NA, Genus=NA)
taxax = rbind(taxax, Other)
rownames(taxax)[7] = "Other"
top_Phylum = as.character(taxax$Phylum)
taxax = as.matrix(taxax)
samdf = sample_data(x2)

## Create Phyloseq Object
ps_x = phyloseq(otu_table(x,taxa_are_rows = F),sample_data(samdf), tax_table(taxax))

## Group by Samdf Category
#gp.norm_m =  merge_samples(ps_x, "NTM_Counts_Categories")
gp.norm_m =  merge_samples(ps_x, "NTM_culture")
sample_data(gp.norm_m)$NTM_culture <- factor(sample_names(gp.norm_m))

## Normalize Counts
ps_x2 = transform_sample_counts(gp.norm_m, function(x) x / sum(x) )
```

##### Culture
```{r fig.width=6, fig.height=8, echo=FALSE}
png('../Images/SupplementaryFigure2C.png', units="in", width=6, height=8, res=600)
p = plot_bar(ps_x2, fill = "Phylum")
p + theme_bw()+ geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = top_Phylum)+ labs(title="Phylum Plot by Mycobacterium Culture Status")+ theme_bw() + theme(axis.text=element_text(size=18),axis.title=element_text(size=18), axis.text.x = element_text(angle = 0, hjust = .5, size = 16, face = "bold")) + xlab("") + ylab("Relative Abundance") + scale_x_discrete(labels= c(" Negative \nCulture", " Positive \nCulture"))
dev.off()
```

##### Count
```{r}
## Group by Samdf Category
gp.norm_m =  merge_samples(ps_x, "NTM_Counts_Categories")
sample_data(gp.norm_m)$NTM_Count_Categories <- factor(sample_names(gp.norm_m))

## Normalize Counts
ps_x2 = transform_sample_counts(gp.norm_m, function(x) x / sum(x) )
```

```{r fig.width=6, fig.height=8, echo=FALSE}
# png('../Images/TaxTablebyMycoCount.png', units="in", width=6, height=8, res=600)
# p = plot_bar(ps_x2, fill = "Phylum")
# p + theme_bw()+ geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = top_Phylum)+ labs(title="Phylum Plot by Mycobacterium Count Category")+ theme_bw() + theme(axis.text=element_text(size=18),axis.title=element_text(size=18), axis.text.x = element_text(angle = 0, hjust = .5, size = 16, face = "bold")) + xlab("") + ylab("Relative Abundance") + scale_x_discrete(labels= c("High \nCounts", "Low \nCounts"))
# dev.off()
```


#### Family Biodiversity 
```{r warning=FALSE, echo=FALSE}
b2 = tax_glom(ps_n, taxrank = "Family")
Family10 = names(sort(taxa_sums(b2), TRUE)[1:7])
x2 = prune_taxa(Family10, b2)

## Add Grouped Column
x = as.data.frame(otu_table(x2))
x$Other = 1 - rowSums(x)
x = as.matrix(x)
taxax = as.data.frame(tax_table(x2))
Other = data.frame(Kingdom="Other", Phylum="Other", Class="NA", Order="NA", Family="Other", Genus="Other")
taxax = rbind(taxax, Other)
rownames(taxax)[8] = "Other"
top_Genus = as.character(taxax$Family)
taxax = as.matrix(taxax)
samdf = sample_data(x2)

## Create Phyloseq Object
ps_x = phyloseq(otu_table(x,taxa_are_rows = F),sample_data(samdf), tax_table(taxax))

## Group by Samdf Category
gp.norm_m =  merge_samples(ps_x, "NTM_culture")
sample_data(gp.norm_m)$NTM_culture <- factor(sample_names(gp.norm_m))

## Normalize Counts
ps_x2 = transform_sample_counts(gp.norm_m, function(x) x / sum(x) )
```

```{r fig.width=6, fig.height=8, echo=FALSE}
p = plot_bar(ps_x2, fill = "Family")
p + geom_bar(aes(color=Family, fill=Family), stat="identity", position="stack", show.legend = F)+ scale_fill_discrete(breaks = top_Genus)+ labs(title="Family Plot by Mycobacterium Culture Status")+theme(axis.text=element_text(size=18),axis.title=element_text(size=18,face="bold"), axis.text.x = element_text(angle = 0, hjust = .5), axis.text.y=element_blank(), axis.ticks.y=element_blank()) + xlab("Mycobacterium Culture Status") + scale_x_discrete(labels= c("Negative", "Positive"))
```


